core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala

```diff
@@ -475,6 +533,194 @@ object ConsumerGroupCommand extends Logging {
+      val groupId = opts.options.valueOf(opts.groupOpt)
+      val (state, assignments) = collectGroupAssignment(groupId)
+      assignments match {
+        case None =>
+          // applies to both old and new consumer
+          printError(s"The consumer group '$groupId' does not exist.")
+          Map.empty
+        case Some(assignments) =>
+          state match {
+            case Some("Dead") =>
+              printError(s"Consumer group '$groupId' does not exist.")
+              Map.empty
+            case Some("Empty") =>
+              val assignmentsToReset = getAssignmentsToReset(assignments)
+              val assignmentsPrepared = prepareAssignmentsToReset(assignmentsToReset)
+              val execute = opts.options.has(opts.executeOpt)
+              if(execute)
+                resetAssignments(assignmentsPrepared)
+              assignmentsPrepared
+            case Some("PreparingRebalance") | Some("AwaitingSync") =>
+              printError(s"Consumer group '$groupId' offsets cannot be reset if it is rebalancing.")
+              Map.empty
+            case Some("Stable") =>
+              printError(s"Consumer group '$groupId' offsets cannot be reset if it has members active.")
+              Map.empty
+            case other =>
+              // the control should never reach here
+              throw new KafkaException(s"Expected a valid consumer group state, but found '${other.getOrElse("NONE")}'.")
+          }
+      }
+    }
+
+    def getAssignmentsToReset(assignments: Seq[PartitionAssignmentState]): Seq[PartitionAssignmentState] = {
+      if (opts.options.has(opts.allTopicsOpt)) {
+        assignments
+      } else if(opts.options.has(opts.topicOpt)) {
+        val topics = opts.options.valuesOf(opts.topicOpt).asScala
+        val topicsPartitions: Map[String, List[Long]] = topics.map {
+          case t if t.contains(":") => t.split(":")(0) -> t.split(":")(1).split(",").map(p => p.toLong).toList
+          case t => t -> Nil
+        }.toMap
+        assignments.filter(assignment => topicsPartitions.contains(assignment.topic.get))
+                   .filter(assignment => {
+                     topicsPartitions(assignment.topic.get) match {
+                       case Nil => true
+                       case partitions => partitions.contains(assignment.partition.get)
+                     }
+                   })
+      } else {
+        CommandLineUtils.printUsageAndDie(opts.parser, "One of the reset scopes should be defined: --all-topics, --topic.")
+      }
+    }
+
+    def parseResetPlan(resetPlanCsv: String): Map[TopicPartition, Long] = {
+      resetPlanCsv.split("\n").map(line => {
+        val Array(topic, partition, offset) = line.split(",").map(_.trim)
+        new TopicPartition(topic, partition.asInstanceOf[Int]) -> offset.asInstanceOf[Long]
+      }).toMap
+    }
+
+    def prepareAssignmentsToReset(assignmentsToReset: Seq[PartitionAssignmentState]): Map[PartitionAssignmentState, Long] = {
```

Comment:
Same as comment above. I think we should try to make offsets the unit we're working with, not assignments. In other words, we should try to avoid this overloaded usage of `PartitionAssignmentState`. If we do so, we'll eliminate a lot of the awkwardness having to deal with all the optional values.